subversion/svnrdump/svnrdump.1
.\"
.\"
.\"     Licensed to the Apache Software Foundation (ASF) under one
.\"     or more contributor license agreements.  See the NOTICE file
.\"     distributed with this work for additional information
.\"     regarding copyright ownership.  The ASF licenses this file
.\"     to you under the Apache License, Version 2.0 (the
.\"     "License"); you may not use this file except in compliance
.\"     with the License.  You may obtain a copy of the License at
.\"
.\"       http://www.apache.org/licenses/LICENSE-2.0
.\"
.\"     Unless required by applicable law or agreed to in writing,
.\"     software distributed under the License is distributed on an
.\"     "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.\"     KIND, either express or implied.  See the License for the
.\"     specific language governing permissions and limitations
.\"     under the License.
.\"
.\"
.\" You can view this file with:
.\" nroff -man [filename]
.\"
.TH svnrdump 1
.SH NAME
svnrdump \- Subversion remote repository dumper and loader
.SH SYNOPSIS
.TP
\fBsvnrdump\fP \fIcommand\fP \fIurl\fP [\fIoptions\fP] [\fIargs\fP]
.SH OVERVIEW
Subversion is a version control system, which allows you to keep old
versions of files and directories (usually source code), keep a log of
who, when, and why changes occurred, etc., like CVS, RCS or SCCS.
\fBSubversion\fP keeps a single copy of the master sources.  This copy
is called the source ``repository''; it contains all the information
to permit extracting previous versions of those files at any time.

For more information about the Subversion project, visit
http://subversion.apache.org.

Documentation for Subversion and its tools, including detailed usage
explanations of the \fBsvn\fP, \fBsvnadmin\fP, \fBsvnserve\fP and
\fBsvnlook\fP programs, historical background, philosophical
approaches and reasonings, etc., can be found at
http://svnbook.red-bean.com/.

Run `svnrdump help' to access the built-in tool documentation.


CHANGES
Version 1.6.12
(21 Jun 2010, from /branches/1.6.x)
http://svn.apache.org/repos/asf/subversion/tags/1.6.12

  User-visible changes:
   * further improvements for issue #3242
   * allow deletion of uris which need character escaping (issue #3636)
   * fix errors with 'svn mkdir --parents' (issue #3649)
   * update address to which crash reports are sent (r901304)
   * check for server certificate revocation on Windows (r898048)
   * disable custom file mutexes on Windows (r879902, -16)
   * fix handling of peg revision'd copy targets (issue #3651)
   * more improvements to 'svn merge --reintegrate' (r935631)
   * allow copying of broken symlinks (issue #3303)
   * improve rep-sharing performance on high-concurrency repos (issue #3506)
   * fixed: added subtrees with mergeinfo break reintegrate (issue #3654)
   * fixed: assertion triggered by tree-conflicted externals (issue #3469)

  Developer-visible changes:
   * give windows devs more flexibility with sqlite versions (r944635)
   * allow the pack tests to work with low file descriptor limits (r937610)
   * improve exception handling on Windows Vista and 7 (r878447, -910, -916)


Version 1.6.11
(19 Apr 2010, from /branches/1.6.x)
http://svn.apache.org/repos/asf/subversion/tags/1.6.11

 User-visible changes:
  * fix for repositories mounted via NFS (issue #3501)
  * enable TCP keep-alives in svnserve (r880552)
  * tighten restrictions on revprops for 'svnadmin verify' (r904594)
  * make ra_serf give better out-of-date information (issue #3561)
  * improve error message upon connection failure with svn+ssh:// (r922516)
  * allow 'svn log' on an uncommitted copy/move destination (r901752)
  * make 'svnadmin hotcopy' copy the fsfs config file (r905303)
  * mergeinfo improvements with non-inheritable mergeinfo (issue #3573)
  * make mergeinfo queries not require access to the repo root (issue #3242)
  * update URLs to refer the new apache.org repository (r904301, -94)
  * update relative externals during a switch (issue #3390)
  * fix 'merge --reintegrate' with self-referential mergeinfo (r892050, -85)
  * improve wc-ng working copy detection (r929382)
  * improve handling of mergeinfo when using serf (r880461)
  * fixed: 'svnlook plist --revprop' with '-t TXN_NAME' (r917640, -8211)
  * fixed: file external from URL cannot overwrite existing item (issue #3552)
  * fixed: potential memory error in 'svn status' (r923674, -9)
  * fixed: merge records mergeinfo from natural history gaps (issue #3432)
  * fixed: theoretical possibility of DB corruption (r926151, -67)

 Developer-visible changes:
  * disable checks for wc-ng working copies when running the test suite
  * on Windows, don't ignore move operation error codes (r896915)
  * more precise reporting of errors occuring with sqlite init (r927323, -8)
  * ensure rangelist APIs are commutative (r923389, -91)


Version 1.6.10
(Not released, see changes for 1.6.11.)


Version 1.6.9
(25 Jan 2010, from /branches/1.6.x)
http://svn.apache.org/repos/asf/subversion/tags/1.6.9

 User-visible changes:
  * allow multiple external updates over ra_svn (issue #3487)
  * fix a segmentation fault when using FSFS (r881905)
  * support Berkeley DB 4.8 (r879688)
  * various autoprop improvements (r880274, -5)
  * improve usage of svn+ssh:// on Windows (issue #2580)
  * teach 1.6.x to recognize 1.7 working copies (1.6.x-future-proof branch)
  * update help text for 'svn update' and 'svn switch' (r886164, -97)
  * make 'svnadmin load --parent-dir' create valid mergeinfo (r888979, -9081)
  * tolerate relative merge source paths in mergeinfo (r889840)
  * teach mod_dav_svn to support the Label header (issue #3519)
  * fixed: svnsync leaves stale sync-locks on mirrors (r884842)
  * fix applicability of 'svn resolve --accept=theirs-conflict' (r880525, -6)
  * fixed: segfault in 'svn resolve' (r896522, -47)
  * fix commit failure against an out-of-date mirror (r900797)

 Developer-visible changes:
  * update ruby bindings test expectation (r880162)
  * don't allow rangelist and mergeinfo API to modify input args (r879093)


Version 1.6.8
(Not released, see changes for 1.6.9.)


Version 1.6.7
(Not released, see changes for 1.6.9.)


[ Note: All revision numbers for versions prior to 1.6.7 reference the
  original repository on svn.collab.net.  For more information see:
  http://svn.apache.org/repos/asf/subversion/README ]


Version 1.6.6
(22 Oct 2009, from /branches/1.6.x)
http://svn.apache.org/repos/asf/subversion/tags/1.6.6

 User-visible changes:
  * fix crash during 'svn update' (r39673)
  * respect Apache's ServerSignature directive (r40008, -21, -31)
  * don't add a file with mixed line endings, and then abort (issue #2713)
  * support Neon 0.29.
  * fix a crash in 'svn rm --force' (r37953)
  * handle tree conflicts involving replacements (issue #3486)
  * allow non-threadsafe sqlite if APR has no threads (r39301)

subversion/svnrdump/svnrdump.c
/*
 *  svnrdump.c: Produce a dumpfile of a local or remote repository
 *  without touching the filesystem, but for temporary files.
 *
 * ====================================================================
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 * ====================================================================
 */

#include "svn_pools.h"
#include "svn_cmdline.h"
#include "svn_client.h"
#include "svn_hash.h"
#include "svn_ra.h"
#include "svn_repos.h"
#include "svn_path.h"
#include "svn_utf.h"
#include "svn_private_config.h"
#include "svn_string.h"
#include "svn_props.h"

#include "dump_editor.h"
#include "load_editor.h"

#include "private/svn_cmdline_private.h"

static svn_opt_subcommand_t dump_cmd, load_cmd;

enum svn_svnrdump__longopt_t
  {
    opt_config_dir = SVN_OPT_FIRST_LONGOPT_ID,
    opt_auth_username,
    opt_auth_password,
    opt_non_interactive,
    opt_auth_nocache,
    opt_version,
    opt_config_option,
  };

static const svn_opt_subcommand_desc2_t svnrdump__cmd_table[] =
  {
    { "dump", dump_cmd, { 0 },
      N_("usage: svnrdump dump URL [-r LOWER[:UPPER]]\n\n"
         "Dump revisions LOWER to UPPER of repository at remote URL "
         "to stdout in a 'dumpfile' portable format.\n"
         "If omitted, LOWER defaults to zero and UPPER to the latest "
         "latest revision.\n"),
      { 0 } },
    { "load", load_cmd, { 0 },
      N_("usage: svnrdump load URL\n\n"
         "Load a 'dumpfile' given on stdin to a repository "
         "at remote URL.\n"),
      { 0 } },
    { "help", 0, { "?", "h" },
      N_("usage: svnrdump help [SUBCOMMAND...]\n\n"
         "Describe the usage of this program or its subcommands.\n"),
      { 0 } },
    { NULL, NULL, { 0 }, NULL, { 0 } }
  };

static const apr_getopt_option_t svnrdump__options[] =
  {
    {"revision",     'r', 1, N_("REV1[:REV2] range of revisions to dump")},
    {"quiet",         'q', 0, N_("no progress (only errors) to stderr")},
    {"config-dir",    opt_config_dir, 1, N_("read user configuration files from"
                                            " directory ARG") },
    {"username",      opt_auth_username, 1, N_("specify a username ARG")},
    {"password",      opt_auth_password, 1, N_("specify a password ARG")},
    {"non-interactive", opt_non_interactive, 0, N_("do no interactive"
                                                   " prompting")},
    {"no-auth-cache", opt_auth_nocache, 0, N_("do not cache authentication"
                                              " tokens")},

    {"help",          'h', 0, N_("display this help")},
    {"version",       opt_version, 0, N_("show program version information")},
    {"config-option", opt_config_option, 1,
                    N_("set user configuration option in the format:\n"
                       "                             "
                       "    FILE:SECTION:OPTION=[VALUE]\n"
                       "                             "
                       "For example:\n"
                       "                             "
                       "    servers:global:http-library=serf")},
    {0,                  0,   0, 0}
  };

/* Baton for the RA replay session. */
struct replay_baton {
  /* The editor producing diffs. */
  const svn_delta_editor_t *editor;

  /* Baton for the editor. */
  void *edit_baton;

  /* Whether to be quiet. */
  svn_boolean_t quiet;
};

typedef struct {
  svn_ra_session_t *session;
  const char *url;
  svn_revnum_t start_revision;
  svn_revnum_t end_revision;
  svn_boolean_t quiet;
} opt_baton_t;

static svn_error_t *
replay_revstart(svn_revnum_t revision,
                void *replay_baton,
                const svn_delta_editor_t **editor,
                void **edit_baton,
                apr_hash_t *rev_props,
                apr_pool_t *pool)
{
  struct replay_baton *rb = replay_baton;
  svn_stringbuf_t *propstring;
  svn_stream_t *stdout_stream;
  svn_stream_t *revprop_stream;

  svn_stream_for_stdout(&stdout_stream, pool);

  /* Revision-number: 19 */
  SVN_ERR(svn_stream_printf(stdout_stream, pool,
                            SVN_REPOS_DUMPFILE_REVISION_NUMBER
                            ": %ld\n", revision));
  propstring = svn_stringbuf_create_ensure(0, pool);
  revprop_stream = svn_stream_from_stringbuf(propstring, pool);
  SVN_ERR(svn_hash_write2(rev_props, revprop_stream, "PROPS-END", pool));
  SVN_ERR(svn_stream_close(revprop_stream));

  /* Prop-content-length: 13 */
  SVN_ERR(svn_stream_printf(stdout_stream, pool,
                            SVN_REPOS_DUMPFILE_PROP_CONTENT_LENGTH
                            ": %" APR_SIZE_T_FMT "\n", propstring->len));

  /* Content-length: 29 */
  SVN_ERR(svn_stream_printf(stdout_stream, pool,
                            SVN_REPOS_DUMPFILE_CONTENT_LENGTH
                            ": %" APR_SIZE_T_FMT "\n\n", propstring->len));

  /* Property data. */
  SVN_ERR(svn_stream_write(stdout_stream, propstring->data,
                           &(propstring->len)));

  SVN_ERR(svn_stream_printf(stdout_stream, pool, "\n"));
  SVN_ERR(svn_stream_close(stdout_stream));

  /* Extract editor and editor_baton from the replay_baton and
     set them so that the editor callbacks can use them. */
  *editor = rb->editor;
  *edit_baton = rb->edit_baton;

  return SVN_NO_ERROR;
}

static svn_error_t *
replay_revend(svn_revnum_t revision,
              void *replay_baton,
              const svn_delta_editor_t *editor,
              void *edit_baton,
              apr_hash_t *rev_props,
              apr_pool_t *pool)
{
  /* No resources left to free. */
  struct replay_baton *rb = replay_baton;
  if (! rb->quiet)
    svn_cmdline_fprintf(stderr, pool, "* Dumped revision %lu.\n", revision);
  return SVN_NO_ERROR;
}

/* Return in *SESSION a new RA session to URL.
 * Allocate *SESSION and related data structures in POOL.
 * Use CONFIG_DIR and pass USERNAME, PASSWORD, CONFIG_DIR and
 * NO_AUTH_CACHE to initialize the authorization baton.
 * CONFIG_OPTIONS (if not NULL) is a list of configuration overrides. */
static svn_error_t *
open_connection(svn_ra_session_t **session,
                const char *url,
                svn_boolean_t non_interactive,
                const char *username,
                const char *password,
                const char *config_dir,
                svn_boolean_t no_auth_cache,
                apr_array_header_t *config_options,
                apr_pool_t *pool)
{
  svn_client_ctx_t *ctx = NULL;
  svn_config_t *cfg_config;

  SVN_ERR(svn_ra_initialize(pool));

  SVN_ERR(svn_config_ensure(config_dir, pool));
  SVN_ERR(svn_client_create_context(&ctx, pool));

  SVN_ERR(svn_config_get_config(&(ctx->config), config_dir, pool));

  if (config_options)
    SVN_ERR(svn_cmdline__apply_config_options(ctx->config, config_options,
                                              "svnrdump: ", "--config-option"));

  cfg_config = apr_hash_get(ctx->config, SVN_CONFIG_CATEGORY_CONFIG,
                            APR_HASH_KEY_STRING);

  /* Default authentication providers for non-interactive use */
  SVN_ERR(svn_cmdline_create_auth_baton(&(ctx->auth_baton), non_interactive,
                                        username, password, config_dir,
                                        no_auth_cache, FALSE, cfg_config,
                                        ctx->cancel_func, ctx->cancel_baton,
                                        pool));
  SVN_ERR(svn_client_open_ra_session(session, url, ctx, pool));
  return SVN_NO_ERROR;
}

static svn_error_t *
replay_revisions(svn_ra_session_t *session, const char *url,
                 svn_revnum_t start_revision, svn_revnum_t end_revision,
                 svn_boolean_t quiet, apr_pool_t *pool)
{
  const svn_delta_editor_t *dump_editor;
  struct replay_baton *replay_baton;
  void *dump_baton;
  const char *uuid;
  svn_stream_t *stdout_stream;

  SVN_ERR(svn_stream_for_stdout(&stdout_stream, pool));

  SVN_ERR(get_dump_editor(&dump_editor, &dump_baton, stdout_stream, pool));

  replay_baton = apr_pcalloc(pool, sizeof(*replay_baton));
  replay_baton->editor = dump_editor;
  replay_baton->edit_baton = dump_baton;
  replay_baton->quiet = quiet;

  /* Write the magic header and UUID */
  SVN_ERR(svn_stream_printf(stdout_stream, pool,
                            SVN_REPOS_DUMPFILE_MAGIC_HEADER ": %d\n\n",
                            SVN_REPOS_DUMPFILE_FORMAT_VERSION));
  SVN_ERR(svn_ra_get_uuid2(session, &uuid, pool));
  SVN_ERR(svn_stream_printf(stdout_stream, pool,
                            SVN_REPOS_DUMPFILE_UUID ": %s\n\n", uuid));

  /* Fake revision 0 if necessary */
  if (start_revision == 0)
    {
      apr_hash_t *prophash;
      svn_stringbuf_t *propstring;
      svn_stream_t *propstream;
      SVN_ERR(svn_stream_printf(stdout_stream, pool,
                                SVN_REPOS_DUMPFILE_REVISION_NUMBER
                                ": %ld\n", start_revision));

      prophash = apr_hash_make(pool);
      propstring = svn_stringbuf_create("", pool);

      SVN_ERR(svn_ra_rev_proplist(session, start_revision,
                                  &prophash, pool));

      propstream = svn_stream_from_stringbuf(propstring, pool);
      SVN_ERR(svn_hash_write2(prophash, propstream, "PROPS-END", pool));
      SVN_ERR(svn_stream_close(propstream));

      /* Property-content-length: 14; Content-length: 14 */
      SVN_ERR(svn_stream_printf(stdout_stream, pool,
                                SVN_REPOS_DUMPFILE_PROP_CONTENT_LENGTH
                                ": %" APR_SIZE_T_FMT "\n",
                                propstring->len));
      SVN_ERR(svn_stream_printf(stdout_stream, pool,
                                SVN_REPOS_DUMPFILE_CONTENT_LENGTH
                                ": %" APR_SIZE_T_FMT "\n\n",
                                propstring->len));
      /* The properties */
      SVN_ERR(svn_stream_write(stdout_stream, propstring->data,
                               &(propstring->len)));
      SVN_ERR(svn_stream_printf(stdout_stream, pool, "\n"));
      if (! quiet)
        svn_cmdline_fprintf(stderr, pool, "* Dumped revision %lu.\n", start_revision);

      start_revision++;
    }

  SVN_ERR(svn_ra_replay_range(session, start_revision, end_revision,
                              0, TRUE, replay_revstart, replay_revend,
                              replay_baton, pool));
  SVN_ERR(svn_stream_close(stdout_stream));
  return SVN_NO_ERROR;
}

static svn_error_t *
load_revisions(svn_ra_session_t *session, const char *url,
               svn_boolean_t quiet, apr_pool_t *pool)
{
  apr_file_t *stdin_file;
  svn_stream_t *stdin_stream;
  const svn_repos_parse_fns2_t *parser;
  void *parse_baton;

  apr_file_open_stdin(&stdin_file, pool);
  stdin_stream = svn_stream_from_aprfile2(stdin_file, FALSE, pool);

  SVN_ERR(get_dumpstream_loader(&parser, &parse_baton, session, pool));
  SVN_ERR(drive_dumpstream_loader(stdin_stream, parser, parse_baton, session, pool));

  svn_stream_close(stdin_stream);

  return SVN_NO_ERROR;
}


static const char *
ensure_appname(const char *progname, apr_pool_t *pool)
{
  if (!progname)
    return "svnrdump";

  progname = svn_dirent_internal_style(progname, pool);
  return svn_dirent_basename(progname, NULL);
}

static svn_error_t *
usage(const char *progname, apr_pool_t *pool)
{
  progname = ensure_appname(progname, pool);

  SVN_ERR(svn_cmdline_fprintf(stderr, pool,
                              _("Type '%s help' for usage.\n"),
                              progname));
  return SVN_NO_ERROR;
}

static svn_error_t *
version(const char *progname, apr_pool_t *pool)
{
  const char *ra_desc_start
    = _("The following repository access (RA) modules are available:\n\n");

  svn_stringbuf_t *version_footer = svn_stringbuf_create(ra_desc_start,
                                                         pool);

  SVN_ERR(svn_ra_print_modules(version_footer, pool));

  progname = ensure_appname(progname, pool);

  return svn_opt_print_help3(NULL, progname, TRUE, FALSE,
                             version_footer->data, NULL, NULL,
                             NULL, NULL, NULL, pool);
}


/** A statement macro, similar to @c SVN_ERR, but returns an integer.
 *
 * Evaluate @a expr. If it yields an error, handle that error and
 * return @c EXIT_FAILURE.
 */
#define SVNRDUMP_ERR(expr)                                              \
  do                                                                    \
    {                                                                   \
      svn_error_t *svn_err__temp = (expr);                              \
      if (svn_err__temp)                                                \
        {                                                               \
          svn_handle_error2(svn_err__temp, stderr, FALSE, "svnrdump: "); \
          svn_error_clear(svn_err__temp);                               \
          return EXIT_FAILURE;                                          \
        }                                                               \
    }                                                                   \
  while (0)

static svn_error_t *
dump_cmd(apr_getopt_t *os, void *baton, apr_pool_t *pool)
{
  opt_baton_t *opt_baton = baton;
  SVN_ERR(replay_revisions(opt_baton->session, opt_baton->url,
                           opt_baton->start_revision, opt_baton->end_revision,
                           opt_baton->quiet, pool));
  return SVN_NO_ERROR;
}

static svn_error_t *
load_cmd(apr_getopt_t *os, void *baton, apr_pool_t *pool)
{
  opt_baton_t *opt_baton = baton;
  SVN_ERR(load_revisions(opt_baton->session, opt_baton->url,
                         opt_baton->quiet, pool));
  return SVN_NO_ERROR;
}

static svn_error_t *
help_cmd(apr_getopt_t *os, void *baton, apr_pool_t *pool)
{
  const char *header =
    _("general usage: svnrdump SUBCOMMAND URL [-r LOWER[:UPPER]]\n"
      "Type 'svnrdump help <subcommand>' for help on a specific subcommand.\n\n"
      "Available subcommands:\n");

  SVN_ERR(svn_opt_print_help3(os, "svnrdump", FALSE, FALSE, NULL, header,
                              svnrdump__cmd_table, svnrdump__options, NULL,
                              NULL, pool));

  return SVN_NO_ERROR;
}

int
main(int argc, const char **argv)
{
  const svn_opt_subcommand_desc2_t *subcommand = NULL;
  opt_baton_t *opt_baton;
  char *revision_cut = NULL;
  svn_revnum_t latest_revision = svn_opt_revision_unspecified;
  apr_pool_t *pool = NULL;
  const char *config_dir = NULL;
  const char *username = NULL;
  const char *password = NULL;
  svn_boolean_t no_auth_cache = FALSE;
  svn_boolean_t non_interactive = FALSE;
  apr_array_header_t *config_options = NULL;
  apr_getopt_t *os;
  const char *first_arg;

  if (svn_cmdline_init ("svnrdump", stderr) != EXIT_SUCCESS)
    return EXIT_FAILURE;

  pool = svn_pool_create(NULL);
  opt_baton = apr_pcalloc(pool, sizeof(*opt_baton));
  opt_baton->start_revision = svn_opt_revision_unspecified;
  opt_baton->end_revision = svn_opt_revision_unspecified;
  opt_baton->url = NULL;

  SVNRDUMP_ERR(svn_cmdline__getopt_init(&os, argc, argv, pool));

  os->interleave = TRUE; /* Options and arguments can be interleaved */

  while (1)
    {
      int opt;
      const char *opt_arg;
      apr_status_t status = apr_getopt_long(os, svnrdump__options, &opt,
                                            &opt_arg);

      if (APR_STATUS_IS_EOF(status))
        break;
      if (status != APR_SUCCESS)
        {
          SVNRDUMP_ERR(usage(argv[0], pool));
          exit(EXIT_FAILURE);
        }

      switch(opt)
        {
        case 'r':
          {
            revision_cut = strchr(opt_arg, ':');
            if (revision_cut)
              {
                opt_baton->start_revision = (svn_revnum_t)strtoul(opt_arg,
                                                                  &revision_cut, 10);
                opt_baton->end_revision = (svn_revnum_t)strtoul(revision_cut + 1,
                                                                NULL, 10);
              }
            else
              opt_baton->start_revision = (svn_revnum_t)strtoul(opt_arg, NULL, 10);
          }
          break;
        case 'q':
          opt_baton->quiet = TRUE;
          break;
        case opt_config_dir:
          config_dir = opt_arg;
          break;
        case opt_version:
          SVNRDUMP_ERR(version(argv[0], pool));
          exit(EXIT_SUCCESS);
          break;
        case 'h':
          SVNRDUMP_ERR(help_cmd(os, opt_baton, pool));
          exit(EXIT_SUCCESS);
          break;
        case opt_auth_username:
          SVNRDUMP_ERR(svn_utf_cstring_to_utf8(&username, opt_arg, pool));
          break;
        case opt_auth_password:
          SVNRDUMP_ERR(svn_utf_cstring_to_utf8(&password, opt_arg, pool));
          break;
        case opt_auth_nocache:
          no_auth_cache = TRUE;
          break;
        case opt_non_interactive:
          non_interactive = TRUE;
          break;
        case opt_config_option:
          if (!config_options)
              config_options =
                    apr_array_make(pool, 1,
                                   sizeof(svn_cmdline__config_argument_t*));

            SVNRDUMP_ERR(svn_utf_cstring_to_utf8(&opt_arg, opt_arg, pool));
            SVNRDUMP_ERR(svn_cmdline__parse_config_option(config_options,
                                                          opt_arg, pool));
        }
    }

  if (os->ind >= os->argc)
    {
      svn_error_clear
                (svn_cmdline_fprintf(stderr, pool,
                                     _("Subcommand argument required\n")));
      SVNRDUMP_ERR(help_cmd(NULL, NULL, pool));
      svn_pool_destroy(pool);
      exit(EXIT_FAILURE);
    }

  first_arg = os->argv[os->ind++];

  subcommand = svn_opt_get_canonical_subcommand2(svnrdump__cmd_table,
                                                 first_arg);

  if (subcommand == NULL)
    {
      const char *first_arg_utf8;
      svn_error_t *err = svn_utf_cstring_to_utf8(&first_arg_utf8,
                                                 first_arg, pool);
      if (err)
        return svn_cmdline_handle_exit_error(err, pool, "svnrdump: ");
      svn_error_clear(svn_cmdline_fprintf(stderr, pool,
                                          _("Unknown command: '%s'\n"),
                                          first_arg_utf8));
      SVNRDUMP_ERR(help_cmd(NULL, NULL, pool));
      svn_pool_destroy(pool);
      exit(EXIT_FAILURE);
    }

  if (subcommand && strcmp(subcommand->name, "help") == 0)
    {
      SVNRDUMP_ERR(help_cmd(os, opt_baton, pool));
      exit(EXIT_SUCCESS);
    }

  /* Only continue if the only not option argument is a url */
  if ((os->ind != os->argc-1)
      || !svn_path_is_url(os->argv[os->ind]))
    {
      SVNRDUMP_ERR(usage(argv[0], pool));
      exit(EXIT_FAILURE);
    }

  SVNRDUMP_ERR(svn_utf_cstring_to_utf8(&(opt_baton->url), os->argv[os->ind], pool));

  opt_baton->url = svn_uri_canonicalize(os->argv[os->ind], pool);

  SVNRDUMP_ERR(open_connection(&(opt_baton->session),
                               opt_baton->url,
                               non_interactive,
                               username,
                               password,
                               config_dir,
                               no_auth_cache,
                               config_options,
                               pool));

  /* Have sane opt_baton->start_revision and end_revision defaults if unspecified */
  SVNRDUMP_ERR(svn_ra_get_latest_revnum(opt_baton->session, &latest_revision, pool));
  if (opt_baton->start_revision == svn_opt_revision_unspecified)
    opt_baton->start_revision = 0;
  if (opt_baton->end_revision == svn_opt_revision_unspecified)
    opt_baton->end_revision = latest_revision;
  if (opt_baton->end_revision > latest_revision)
    {
      SVN_INT_ERR(svn_cmdline_fprintf(stderr, pool,
                                      _("Revision %ld does not exist.\n"),
                                      opt_baton->end_revision));
      exit(EXIT_FAILURE);
    }
  if (opt_baton->end_revision < opt_baton->start_revision)
    {
      SVN_INT_ERR(svn_cmdline_fprintf(stderr, pool,
                                      _("LOWER cannot be greater "
                                        "than UPPER.\n")));
      exit(EXIT_FAILURE);
    }

  /* Dispatch the subcommand */
  SVNRDUMP_ERR((*subcommand->cmd_func)(os, opt_baton, pool));

  svn_pool_destroy(pool);

  return 0;
}
